src/Client.php

```diff
@@ -79,6 +91,170 @@ public function __construct(Transport $transport, LoggerInterface $logger = null
+     * @throws \Vault\Exceptions\ClientException
+     * @throws \RuntimeException
+     * @throws \InvalidArgumentException
+     */
+    public function head($url, array $options = [])
+    {
+        return $this->responseBuilder->build($this->send(new Request('HEAD', $url), $options));
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param array            $options
+     *
+     * @return ResponseInterface
+     *
+     * @throws \Psr\Cache\InvalidArgumentException
+     * @throws \Vault\Exceptions\TransportException
+     * @throws \Vault\Exceptions\ClientException
+     * @throws \Vault\Exceptions\ServerException
+     * @throws \RuntimeException
+     * @throws \InvalidArgumentException
+     */
+    public function send(RequestInterface $request, array $options = [])
+    {
+        $request = $request->withHeader('User-Agent', 'VaultPHP/1.0.0');
+        $request = $request->withHeader('Content-Type', 'application/json');
+
+        if ($this->token) {
+            $request = $request->withHeader('X-Vault-Token', $this->token->getAuth()->getClientToken());
+        }
+
+        $this->logger->debug('Request.', [
+            'method' => $request->getMethod(),
+            'uri' => $request->getUri(),
+            'headers' => $request->getHeaders(),
+            'body' => $request->getBody()->getContents(),
+        ]);
+
+        try {
+            $response = $this->transport->send($request, $options);
+        } catch (TransferException $e) {
+            $this->logger->error('Something went wrong when calling Vault.', [
+                'code' => $e->getCode(),
+                'message' => $e->getMessage(),
+            ]);
+
+            $this->logger->debug('Trace.', ['exception' => $e]);
+
+            throw new ServerException(sprintf('Something went wrong when calling Vault (%s).', $e->getMessage()));
+        }
+
+        $this->logger->debug('Response.', [
+            'statusCode' => $response->getStatusCode(),
+            'reasonPhrase' => $response->getReasonPhrase(),
+            'headers ' => $response->getHeaders(),
+            'body' => $response->getBody()->getContents(),
+        ]);
+
+        // retry request if true
+        if ($this->checkResponse($response)) {
+            return $this->send($request, $options);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Returns true whenever request should be retried.
+     *
+     * @param ResponseInterface $response
+     *
+     * @return bool
+     *
+     * @throws \Vault\Exceptions\ClientException
+     * @throws \Vault\Exceptions\ServerException
+     * @throws \RuntimeException
+     * @throws \Psr\Cache\InvalidArgumentException
+     */
+    protected function checkResponse(ResponseInterface $response)
+    {
+        // start re-authentication process if got 403 code and token is expired
+        if (
+            $response->getStatusCode() === 403 &&
+            !$this->isReAuthenticationInProgress() &&
```

Comment:
obviously :)

but when is it possible? when some request started the reauthentication and some other process was trying to do the same?